concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs

```diff
@@ -0,0 +1,668 @@
+        forM_ (Map.keys nfsn) $
+            \deadUpdate -> transactionTable . ttHashMap . at' (getHash deadUpdate) .= Nothing
+        -- Update the non-finalized chain updates
+        transactionTable
+            . ttNonFinalizedChainUpdates
+            . at' uty
+            ?= (nfcu & (nfcuMap . at' sn .~ Nothing) & (nfcuNextSequenceNumber .~ sn + 1))
+
+-- |Mark a live transaction as committed in a particular block.
+-- This does nothing if the transaction is not live.
+-- A committed transaction cannot be purged while it is committed for a round after the round of
+-- the last finalized block.
+doCommitTransaction ::
+    (MonadState (SkovData pv) m) =>
+    -- |Round of the block
+    Round ->
+    -- |The 'BlockHash' that the transaction should
+    -- be committed to.
+    BlockHash ->
+    -- |The 'TransactionIndex' in the block.
+    TransactionIndex ->
+    -- |The transaction to commit.
+    BlockItem ->
+    m ()
+doCommitTransaction rnd bh ti transaction =
+    transactionTable
+        . ttHashMap
+        . at' (getHash transaction)
+        . traversed
+        . _2
+        %= addResult bh rnd ti
+
+-- |Add a transaction to the transaction table if its nonce/sequence number is at least the next
+-- non-finalized nonce/sequence number. The return value is 'True' if and only if the transaction
+-- was added.
+-- When adding a transaction from a block, use the 'Round' of the block. Otherwise use round @0@.
+-- The transaction must not already be present.
+doAddTransaction :: (MonadState (SkovData pv) m) => Round -> BlockItem -> VerificationResult -> m Bool
+doAddTransaction rnd transaction verRes = do
+    added <- transactionTable %%=! addTransaction transaction (commitPoint rnd) verRes
+    when added $ transactionTablePurgeCounter += 1
+    return added
+
+-- |Mark the provided transaction as dead for the provided 'BlockHash'.
+doMarkTransactionDead ::
+    (MonadState (SkovData pv) m) =>
+    -- |The 'BlockHash' where the transaction was committed.
+    BlockHash ->
+    -- |The 'BlockItem' to mark as dead.
+    BlockItem ->
+    m ()
+doMarkTransactionDead blockHash transaction =
+    transactionTable
+        . ttHashMap
+        . at' (getHash transaction)
+        . mapped
+        . _2
+        %= markDeadResult blockHash
+
+-- |Purge transactions from the transaction table and pending transactions.
+-- Transactions are purged only if they are not included in a live block, and
+-- have either expired or arrived longer ago than the transaction keep alive time.
+--
+-- If the first argument is @False@, the transaction table is only purged if
+-- 'rpInsertionsBeforeTransactionPurged' transactions have been inserted since
+-- the last purge.  If it is true, the table is purged regardless.
+--
+-- WARNING: This function violates the independence of the tree state components.
+-- In particular, the following invariants are assumed and maintained:
+--
```

Comment:
I removed the comment. But I would think that it originally was because it operates both on the pending transactions and the transaction table itself.